tests/testsuite/rustc.rs

```diff
@@ -134,6 +134,263 @@ fn fails_with_args_to_all_binaries() {
+        .masquerade_as_nightly_cargo()
+        .with_status(101)
+        .with_stderr(
+            "[ERROR] the `crate-type` flag is unstable, pass `-Z unstable-options` to enable it
+See https://github.com/rust-lang/cargo/issues/10083 for more information about the `crate-type` flag.",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn fails_with_crate_type_to_multi_binaries() {
+    let p = project()
+        .file("src/bin/foo.rs", "fn main() {}")
+        .file("src/bin/bar.rs", "fn main() {}")
+        .file("src/bin/baz.rs", "fn main() {}")
+        .file("src/lib.rs", r#" "#)
+        .build();
+
+    p.cargo("rustc --crate-type lib -Zunstable-options")
+        .masquerade_as_nightly_cargo()
+        .with_status(101)
+        .with_stderr(
+            "[ERROR] crate types to rustc can only be passed to one target, consider filtering
+the package by passing, e.g., `--lib` or `--example` to specify a single target",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn fails_with_crate_type_to_multi_examples() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name = "foo"
+            version = "0.0.1"
+            authors = []
+
+            [[example]]
+            name = "ex1"
+            crate-type = ["rlib"]
+            [[example]]
+            name = "ex2"
+            crate-type = ["rlib"]
+        "#,
+        )
+        .file("src/lib.rs", "")
+        .file("examples/ex1.rs", "")
+        .file("examples/ex2.rs", "")
+        .build();
+
+    p.cargo("rustc -v --example ex1 --example ex2 --crate-type lib,cdylib -Zunstable-options")
+        .masquerade_as_nightly_cargo()
+        .with_status(101)
+        .with_stderr(
+            "[ERROR] crate types to rustc can only be passed to one target, consider filtering
+the package by passing, e.g., `--lib` or `--example` to specify a single target",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn fails_with_crate_type_to_binary() {
+    let p = project().file("src/bin/foo.rs", "fn main() {}").build();
+
+    p.cargo("rustc --crate-type lib -Zunstable-options")
+        .masquerade_as_nightly_cargo()
+        .with_status(101)
+        .with_stderr(
+            "[ERROR] crate types can only be specified for libraries and example libraries.
+Binaries, tests, and benchmarks are always the `bin` crate type",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn build_with_crate_type_for_foo() {
+    let p = project()
+        .file("src/main.rs", "fn main() {}")
+        .file("src/lib.rs", r#" "#)
+        .build();
+
+    p.cargo("rustc -v --lib --crate-type lib -Zunstable-options")
+        .masquerade_as_nightly_cargo()
+        .with_stderr(
+            "\
+[COMPILING] foo v0.0.1 ([CWD])
+[RUNNING] `rustc --crate-name foo src/lib.rs [..]--crate-type lib \
+        --emit=[..]link[..]-C debuginfo=2 \
+        -C metadata=[..] \
+        --out-dir [..] \
+        -L dependency=[CWD]/target/debug/deps`
```

Comment:
These tests look great!

Just one minor thing, would you mind removing the checks for `-C debuginfo -C metadata --out-dir -L dependency` from these new tests?  They aren't really relevant to what this is trying to validate (the `--crate-type` flag is the important bit).

This helps keep the test a little more succinct and easier to see exactly what it is checking for, and also makes it less of a hassle in case those unrelated flags change in the future.